### PR TITLE
NXOS MDS modules and tests

### PR DESCRIPTION
**Pull request by srbharadwaj**
##### SUMMARY
This PR is for adding new modules for NXOS
Platform Type: MDS switches
Modules added: 
a) nxos_vsan : Configuration of vsan for Cisco MDS NXOS.
b) nxos_devicealias: Configuration of device alias for Cisco MDS NXOS.
c) nxos_zone_zoneset: Configuration of zone/zoneset for Cisco MDS NXOS.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
nxos_vsan, nxos_devicealias, nxos_zone_zoneset

##### ADDITIONAL INFORMATION
This commit contains units tests and integration tests for these modules.
Please note that these modules run only on NXOS MDS switches.
You could find the type of switch via the following command
switch(config)# sh ver | json | i chassis
      <chassis_id>MDS 9710 (10 Slot) Chassis</chassis_id>

### CODE REVIEW

**Comment by srbharadwaj** [comment]
@trishnaguha Can i know when this will be reviewed? Since this is my first PR, i was wondering how much time i might need to allocate for the effort involved in incorporating the review comments.

**Comment by abenokraitis** [comment]
AFAIK `nxos_*` is currently a reserved prefix for modules written and maintained by the Ansible Network Engineering team. Could this be modified to a not-yet-used prefix?

**Comment by srbharadwaj** [comment]
> AFAIK `nxos_*` is currently a reserved prefix for modules written and maintained by the Ansible Network Engineering team. Could this be modified to a not-yet-used prefix?

Can you point me to the personals in the Ansible Network Eng team? I got this PR opened with the help of @mikewiebe

**Comment by mikewiebe** [comment]
@abenokraitis and @trishnaguha These are nxos modules so the should live under the `nxos_` namespace.

File: lib/ansible/modules/network/nxos/nxos_vsan.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+                    - List of vsan's interfaces to be added
+                type: list
+'''
+
+EXAMPLES = '''
+---
+-
+  name: "Test that vsan module works"
+  nxos_vsan:
+    provider: "{{ creds }}"
```
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/nxos_vsan.py]: Please remove `provider` as we don't plan to support connection local for newly added module.

File: lib/ansible/modules/network/nxos/nxos_vsan.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+          - fc1/11
+          - fc1/21
+          - "port-channel 2"
+        name: vsan-SAN-B
+        remove: false
+        suspend: true
+      -
+        id: 1923
+        name: vsan-SAN-Old
+        remove: true
```
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/nxos_vsan.py:76]: The yaml can look like for better readability 
```
      - id: 922
        interface:
          - fc1/1
          - fc1/2
          - "port-channel 1"
        name: vsan-SAN-A
        remove: false
        suspend: false
      - id: 923
        interface:
          - fc1/11
          - fc1/21
          - "port-channel 2"
        name: vsan-SAN-B
        remove: false
        suspend: true
      - id: 1923
        name: vsan-SAN-Old
        remove: true
```

File: lib/ansible/modules/network/nxos/nxos_vsan.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+  sample:
+    - terminal dont-ask
+    - vsan database
+    - vsan 922 interface fc1/40
+    - vsan 922 interface port-channel 155
+    - no terminal dont-ask
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.network.nxos.nxos import load_config, nxos_argument_spec, run_commands
```
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/nxos_vsan.py]: We are not to support connection local for new modules, please remove `nxos_argument_spec` import.

File: lib/ansible/modules/network/nxos/nxos_vsan.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+        remove=dict(type='bool'),
+        suspend=dict(type='bool'),
+        interface=dict(type='list', elements='str')
+    )
+
+    argument_spec = dict(
+        vsan=dict(type='list', elements='dict', options=vsan_element_spec)
+    )
+
+    argument_spec.update(nxos_argument_spec)
```
Comment (trishnaguha) [inline lib/ansible/modules/network/nxos/nxos_vsan.py]: This isn't required for network_cli/httpapi.

File: test/integration/targets/nxos_devicealias/tasks/nxapi.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  set_fact: test_items="{{ test_cases.files | map(attribute='path') | list }}"
+
+- name: run test cases (connection=httpapi)
+  include: "{{ test_case_to_run }} ansible_connection=httpapi connection={{ nxapi }}"
+  with_items: "{{ test_items }}"
+  loop_control:
+    loop_var: test_case_to_run
+
+- name: run test cases (connection=local)
+  include: "{{ test_case_to_run }} ansible_connection=local connection={{ nxapi }}"
```
Comment (trishnaguha) [inline test/integration/targets/nxos_devicealias/tasks/nxapi.yaml:30]: Please remove connection local from tests.

File: test/integration/targets/nxos_devicealias/tests/common/sanity.yaml [hunk 6]
```diff
@@ -0,0 +1,113 @@
+---
+- debug: msg="START connection={{ ansible_connection }} nxos_devicealias sanity test"
+- debug: msg="Using provider={{ connection.transport }}"
+  when: ansible_connection == "local"
```
Comment (trishnaguha) [inline test/integration/targets/nxos_devicealias/tests/common/sanity.yaml]: This isn't required as we are not doing local connection support for new modules.

File: test/integration/targets/nxos_devicealias/tests/common/sanity.yaml [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+  - name: Idempotence Check
+    nxos_devicealias: *config
+    register: result
+
+  - assert: &false
+      that:
+        - "result.changed == false"
+
+  # - name: Configure ntp with some defaults
```
Comment (trishnaguha) [inline test/integration/targets/nxos_devicealias/tests/common/sanity.yaml]: Is there a reason that these testcases are commented out?

File: test/integration/targets/nxos_devicealias/meta/main.yml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+# dependencies:
+#   - prepare_nxos_tests
```
Comment (trishnaguha) [inline test/integration/targets/nxos_devicealias/meta/main.yml]: dependencies can be empty if not required. 
```
dependencies: []
```

File: test/integration/targets/nxos_vsan/meta/main.yml [hunk 9]
```diff
@@ -0,0 +1,2 @@
+# dependencies:
+#   - prepare_nxos_tests
```
Comment (trishnaguha) [inline test/integration/targets/nxos_vsan/meta/main.yml]: dependencies can be empty if not required. 
```
dependencies: []
```

File: test/integration/targets/nxos_vsan/tasks/nxapi.yaml [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  set_fact: test_items="{{ test_cases.files | map(attribute='path') | list }}"
+
+- name: run test cases (connection=httpapi)
+  include: "{{ test_case_to_run }} ansible_connection=httpapi connection={{ nxapi }}"
+  with_items: "{{ test_items }}"
+  loop_control:
+    loop_var: test_case_to_run
+
+- name: run test cases (connection=local)
+  include: "{{ test_case_to_run }} ansible_connection=local connection={{ nxapi }}"
```
Comment (trishnaguha) [inline test/integration/targets/nxos_vsan/tasks/nxapi.yaml:30]: This isn't required anymore.

File: test/integration/targets/nxos_vsan/tests/common/sanity.yaml [hunk 11]
```diff
@@ -0,0 +1,47 @@
+---
+- debug: msg="START nxos_vsan sanity test with connection={{ ansible_connection }} "
+- debug: msg="Using provider={{ connection.transport }}"
+  when: ansible_connection == "local"
```
Comment (trishnaguha) [inline test/integration/targets/nxos_vsan/tests/common/sanity.yaml]: Please remove local.

File: test/integration/targets/nxos_zone_zoneset/meta/main.yml [hunk 12]
```diff
@@ -0,0 +1,2 @@
+# dependencies:
+#   - prepare_nxos_tests
```
Comment (trishnaguha) [inline test/integration/targets/nxos_zone_zoneset/meta/main.yml]: dependencies can be empty if not required. 
```
dependencies: []
```

**Review by trishnaguha — CHANGES_REQUESTED** [review]
Remove provider example, nxos_argument_spec import and local connection tests from all the modules.

**Comment by trishnaguha** [comment]
@srbharadwaj @mikewiebe I believe the integration tests can only run on nxos mds switch which means the nxos integration testsuite is going to fail. The integration tests needs to be verify what kind of platform it is before the tests are executed. Or you can have what's been done for nxos_install_os tests https://github.com/ansible/ansible/blob/devel/test/integration/targets/nxos_install_os/tasks/upgrade/main_os_install.yaml

**Comment by srbharadwaj** [comment]
> @srbharadwaj @mikewiebe I believe the integration tests can only run on nxos mds switch which means the nxos integration testsuite is going to fail as we do not have mds switch. The integration tests needs to be verify what kind of platform it is before the tests are executed. Or you can have what's been done for nxos_install_os tests https://github.com/ansible/ansible/blob/devel/test/integration/targets/nxos_install_os/tasks/upgrade/main_os_install.yaml

Thank you @trishnaguha , will make the changes and get back.

**Comment by mikewiebe** [comment]
@trishnaguha I took a look at these failures but they don't make sense.

```
lib/ansible/modules/network/nxos/nxos_devicealias.py:0:0: module-incorrect-version-added: version_added should be '2.10'. Currently 2.1
lib/ansible/modules/network/nxos/nxos_devicealias.py:0:0: nonexistent-parameter-documented: Argument 'provider' is listed in DOCUMENTATION.options, but not accepted by the module argument_spec
```

The `version_added` field has been changed to `2.10` in each module.  Looks like it may be a bug in shippable?

Also, all references to `provider` have been removed so not sure why this is complaining.

Any thoughts?  Thanks!

**Comment by srbharadwaj** [comment]
@trishnaguha can you please help on this?

**Comment by abenokraitis** [comment]
Greetings, sorry for the delay, we've had two back-to-back travel conferences and a public holiday. Expect to hear back from @trishnaguha later this week!

**Comment by abenokraitis** [comment]
Per the conversation with Gerrard at Cisco, their request was to make this start with `mds_*` so please modify the modules to start with that prefix, thanks!

**Comment by srbharadwaj** [comment]
Opened a new PR (https://github.com/ansible/ansible/pull/64269), hence closing this PR

**Pull request closed without merging**